### PR TITLE
JSON Mutation

### PR DESCRIPTION
**Pull request by macneale4**
Add JSON Value mutation operations and tests. These changes do not alter the operation of dolt in anyway yet - that will come in a second PR which updates the JSON_SET procedure, and adds support for the JSON_REPLACE, JSON_INSERT, and JSON_REMOVE procedures at the same time. This is laying the foundation for that work.

### CODE REVIEW

File: sql/types/json_test.go [hunk 1]
```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
...
+	},
+	{
+		desc:      "empty array last index is 0",
+		doc:       `[]`,
+		path:      "$[last]",
+		value:     `42`,
+		resultVal: `[42]`,
+		changed:   true,
+	},
+
```
Comment (fulghum) [inline sql/types/json_test.go:385]:
```suggestion
```

File: sql/types/json_test.go [hunk 2]
```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
...
+	{
+		desc:      "treating object as an array replaces for index 0",
+		doc:       `{"a":1}`,
+		path:      "$[0]",
+		value:     `42`,
+		resultVal: `42`,
+		changed:   true,
+	},
+
```
Comment (fulghum) [inline sql/types/json_test.go:394]:
```suggestion
```

File: sql/types/json_test.go [hunk 3]
```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
...
+	{
+		desc:      "scalar will append as an array for out of bounds",
+		doc:       `17`,
+		path:      "$[51]",
+		value:     `42`,
+		resultVal: `[17, 42]`,
+		changed:   true,
+	},
+	{
+		desc:      "scalar will be overwritten for index 0",
```
Comment (fulghum) [inline sql/types/json_test.go:676]:
```suggestion
		desc:      "scalar will not be overwritten for index 0",
```

File: sql/types/json_test.go [hunk 4]
```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
...
+	{
+		desc:      "scalar will be prefixed when underflow happens",
+		doc:       `17`,
+		path:      "$[last-23]",
+		value:     `42`,
+		resultVal: `[42, 17]`,
+		changed:   true,
+	},
+	{
+		desc:      "object field updated",
```
Comment (fulghum) [inline sql/types/json_test.go:692]:
```suggestion
		desc:      "existing object field not updated",
```

File: sql/types/json_test.go [hunk 5]
```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
...
+		doc:       `{"a": 1}`,
+		path:      "$.a",
+		value:     `42`,
+		resultVal: `{"a": 1}`,
+		changed:   false,
+	},
+	{
+		desc:      "object field set",
```
Comment (fulghum) [inline sql/types/json_test.go:700]:
```suggestion
		desc:      "new object field inserted",
```

File: sql/types/json_test.go [hunk 6]
```diff
@@ -290,3 +290,803 @@ func TestJsonRoundTripping(t *testing.T) {
...
+	}
+}
+
+type parseErrTest struct {
+	desc         string
+	path         string
+	expectErrStr string
+}
+
+var JsonPathParseErrTests = []parseErrTest{
```
Comment (fulghum) [inline sql/types/json_test.go:1022]: Should we add cases for using `*` or `**` wildcards? I noticed several json functions mentioned those are not allowed in the paths.

**Review by fulghum — APPROVED** [review]
Code look correct and the tests look beautiful! 

Just some really small comments. Biggest one was that the `doc` arg in the new functions in json_value.go seemed like it could be simplified a little bit.

File: sql/types/json_test.go (hunk 6 above)
Comment (macneale4) [reply sql/types/json_test.go:1020]: I'm going to handle this in the next PR.

File: sql/types/json_value.go [hunk 7]
```diff
@@ -548,3 +563,343 @@ func jsonObjectDeterministicOrder(a, b map[string]interface{}, inter []string) (
...
+func walkPathAndUpdate(path string, doc interface{}, val interface{}, mode int, cursor *int) (interface{}, bool, *parseErr) {
+	if path[0] == '.' {
+		path = path[1:]
+		*cursor = *cursor + 1
+		strMap, ok := doc.(map[string]interface{})
+		if !ok {
+			// not a map, can't do anything. NoOp
+			return doc, false, nil
+		}
+		return updateObject(path, doc, strMap, val, mode, cursor)
```
[the remark this answers is not in the document]
Comment (macneale4) [reply sql/types/json_value.go:677]: Yep. Fixed.

File: sql/types/json_value.go [hunk 8]
```diff
@@ -548,3 +563,343 @@ func jsonObjectDeterministicOrder(a, b map[string]interface{}, inter []string) (
...
+	if path[0] == '"' {
+		right := strings.Index(path[1:], "\"")
+		if right < 0 {
+			return "", "", &parseErr{msg: "Invalid JSON path expression. '\"' expected", character: *cursor}
+		}
+		name = path[1 : right+1]
+		remainingPath = path[right+2:]
+		*cursor = *cursor + right + 2
+	} else {
+		regex := regexp.MustCompile(`^([a-zA-Z0-9_]+)(.*)$`)
```
[the remark this answers is not in the document]
Comment (macneale4) [reply sql/types/json_value.go]: Fixed

**Pull request merged**